File: snippets/cpp/math/sieve.cpp
vi sieve(int n) {
    vector<bool> is_p(n + 1, true);
    is_p[0] = is_p[1] = false;
    for (int p = 2; p * p <= n; p++)
        if (is_p[p])
            for (int i = p * p; i <= n; i += p)
                is_p[i] = false;
    vi primes;
    for (int p = 2; p <= n; p++)
        if (is_p[p]) primes.pb(p);
    return primes;
}

File: snippets/rust/math/sieve.rs
fn sieve(n: usize) -> Vec<usize> {
    if n < 2 { return Vec::new(); }
    let mut is_p = vec![true; n + 1];
    is_p[0] = false; is_p[1] = false;
    let mut p = 2usize;
    while p * p <= n {
        if is_p[p] {
            let mut i = p * p;
            while i <= n { is_p[i] = false; i += p; }
        }
        p += 1;
    }
    (2..=n).filter(|&x| is_p[x]).collect()
}
